File: CollapseCalendarView/src/main/java/com/wefika/calendar/manager/Month.java
package com.wefika.calendar.manager;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.joda.time.LocalDate;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Blaž Šolar on 24/02/14.
 */
public class Month extends RangeUnit {

    @NotNull private final List<Week> mWeeks = new ArrayList<>();
    private int mSelectedIndex = -1;

    protected Month(@NotNull LocalDate date, @NotNull LocalDate today, @Nullable LocalDate minDate,
                    @Nullable LocalDate maxDate) {
        super(
                date.withDayOfMonth(1),
                date.withDayOfMonth(date.dayOfMonth().getMaximumValue()),
                "MMMM yyyy",
                today,
                minDate,
                maxDate
        );

        build();
    }

    @Override
    public boolean hasNext() {

        LocalDate maxDate = getMaxDate();
        if (maxDate == null) {
            return true;
        } else {

            LocalDate to = getTo();
            int year = maxDate.getYear();
            int yearTo = to.getYear();

            int month = maxDate.getMonthOfYear();
            int monthTo = to.getMonthOfYear();

            return year > yearTo ||
                    (year == yearTo && month > monthTo);

        }
    }

    @Override
    public boolean  hasPrev() {

        LocalDate minDate = getMinDate();
        if (minDate == null) {
            return true;
        } else {

            LocalDate from = getFrom();
            int year = minDate.getYear();
            int yearFrom = from.getYear();

            int month = minDate.getMonthOfYear();
            int monthFrom = from.getMonthOfYear();

            return year < yearFrom ||
                    (year == yearFrom && month < monthFrom);

        }
    }

    @Override
    public boolean next() {

        if (hasNext()) {
            setFrom(getTo().plusDays(1));
            setTo(getFrom().withDayOfMonth(getFrom().dayOfMonth().getMaximumValue()));

            build();
            return true;
        } else {
            return false;
        }
    }

    @Override
    public boolean prev() {

        if (hasPrev()) {
            setFrom(getFrom().minusDays(1).withDayOfMonth(1));
            setTo(getFrom().withDayOfMonth(getFrom().dayOfMonth().getMaximumValue()));

            build();
            return true;
        } else {
            return false;
        }

    }

    @Override
    public void deselect(@NotNull LocalDate date) {
        if (isSelected()) {
            mSelectedIndex = -1;
            for (Week week : mWeeks) {
                setSelected(false);
                week.deselect(date);
            }
        }
    }

    @Override
    public boolean select(@NotNull LocalDate date) {
        int cnt = mWeeks.size();
        for (int i = 0; i < cnt; i++) {
            Week week = mWeeks.get(i);
            if (week.select(date)) {
                mSelectedIndex = i;
                setSelected(true);
                return true;
            }
        }
        return false;
    }

    @NotNull
    public List<Week> getWeeks() {
        return mWeeks;
    }

    public int getSelectedIndex() {
        return mSelectedIndex;
    }

    @Override
    public void build() {

        setSelected(false);
        mWeeks.clear();

        LocalDate date = getFrom().withDayOfWeek(1);
        for (int i = 0; i == 0 || getTo().compareTo(date) >= 0; i++) {
            mWeeks.add(new Week(date, getToday(), getMinDate(), getMaxDate()));
            date = date.plusWeeks(1);
        }

    }

    @Nullable
    @Override
    LocalDate getFirstDateOfCurrentMonth(@NotNull LocalDate currentMonth) {

        int year = currentMonth.getYear();
        int month = currentMonth.getMonthOfYear();

        LocalDate from = getFirstEnabled();
        int fromYear = from.getYear();
        int fromMonth = from.getMonthOfYear();

        if (year == fromYear && month == fromMonth) {
            return from;
        } else {
            return null;
        }

    }
}

File: library/src/main/java/com/github/gfranks/collapsible/calendar/model/Month.kt
package com.github.gfranks.collapsible.calendar.model

import org.joda.time.LocalDate
import java.util.*

class Month(
        date: LocalDate,
        today: LocalDate,
        minDate: LocalDate?,
        maxDate: LocalDate?
) : RangeUnit(date.withDayOfMonth(1), date.withDayOfMonth(date.dayOfMonth().maximumValue), today, minDate, maxDate) {

    private val mWeeks = ArrayList<Week>()
    var selectedIndex = -1
        private set

    override val type = CalendarUnitType.MONTH

    val weeks: List<Week>
        get() = mWeeks

    init {
        build()
    }

    override fun hasNext(): Boolean {
        val maxDate = maxDate ?: return true

        val year = maxDate.year
        val yearTo = dateTo.year

        val month = maxDate.monthOfYear
        val monthTo = dateTo.monthOfYear

        return year > yearTo || year == yearTo && month > monthTo
    }

    override fun hasPrev(): Boolean {
        val minDate = minDate ?: return true

        val from = dateFrom
        val year = minDate.year
        val yearFrom = from.year

        val month = minDate.monthOfYear
        val monthFrom = from.monthOfYear

        return year < yearFrom || year == yearFrom && month < monthFrom
    }

    override fun setPeriod(date: LocalDate): Boolean {
        return if (hasDate(date)) {
            dateFrom = date.withDayOfMonth(1)
            dateTo = dateFrom.withDayOfMonth(dateFrom.dayOfMonth().maximumValue)
            build()
            true
        } else {
            false
        }
    }

    override fun next(): Boolean {
        return if (hasNext()) {
            dateFrom = dateTo.plusDays(1)
            dateTo = dateFrom.withDayOfMonth(dateFrom.dayOfMonth().maximumValue)
            build()
            true
        } else {
            false
        }
    }

    override fun prev(): Boolean {
        return if (hasPrev()) {
            dateFrom = dateFrom.minusDays(1).withDayOfMonth(1)
            dateTo = dateFrom.withDayOfMonth(dateFrom.dayOfMonth().maximumValue)
            build()
            true
        } else {
            false
        }
    }

    override fun deselect(date: LocalDate?) {
        if (date != null && isSelected && isInView(date)) {
            for (week in mWeeks) {
                if (week.isSelected && week.isIn(date)) {
                    selectedIndex = -1
                    isSelected = false
                    week.deselect(date)
                }
            }
        }
    }

    override fun select(date: LocalDate?): Boolean {
        val cnt = mWeeks.size
        for (i in 0 until cnt) {
            val week = mWeeks[i]
            if (week.select(date)) {
                selectedIndex = i
                isSelected = true
                return true
            }
        }
        return false
    }

    override fun build() {
        isSelected = false
        mWeeks.clear()

        var date = dateFrom.withDayOfWeek(1)
        var i = 0
        while (i == 0 || dateTo >= date) {
            mWeeks.add(Week(date, today, minDate, maxDate))
            date = date.plusWeeks(1)
            i++
        }
    }

    override fun getFirstDateOfCurrentMonth(currentMonth: LocalDate?): LocalDate? {
        currentMonth ?: return null

        with(firstEnabled) {
            if (currentMonth.year == year && currentMonth.monthOfYear == monthOfYear) {
                return this
            }
        }

        return null
    }
}
